rr_viz/src/ross/estop/emergency_stop.py

```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/env python
+from PyQt5.QtWidgets import QVBoxLayout, QWidget, QLabel, QPushButton, QHBoxLayout
+from PyQt5.QtGui import QFont
+from PyQt5.QtCore import Qt
+
+import rospy
+from std_srvs.srv import SetBool, SetBoolRequest
+from std_msgs.msg import Bool
+
+class EmergencyStop(QWidget):
+    def __init__(self, parent):
+        super(QWidget, self).__init__(parent)
+        
+        self.layout_init()
+
+        # Set up service
+        self.eStop_srv_name = '/emergency_stop/trigger_eStop'
+        self.eStop_srv = rospy.ServiceProxy(self.eStop_srv_name, SetBool)
+        self.eStop_req = SetBoolRequest()
+        self.eStop_srv_timer = rospy.Timer(rospy.Duration(
+            1), self.ping_srv)
+        self.eStop_srv_connected = False
+
+        # Set up subscriber
+        self.eStop_status_topic_name = '/emergency_stop/status'
+        self.eStop_status_sub = rospy.Subscriber(self.eStop_status_topic_name, Bool, self.eStop_status_update)
+
+    def layout_init(self):
+        self.v_layout = QVBoxLayout()
+
+        # Title
+        self.title_label = QLabel('Emergency Stop')
+        self.title_label.setFont(QFont('Ubuntu', 11, QFont.Bold))
+        self.title_label.setAlignment(Qt.AlignRight)
+        self.v_layout.addWidget(self.title_label)
+        
+        # Buttons
+        self.h_layout = QHBoxLayout()
+        self.enable_eStop_button = QPushButton('Emergency Stop')
+        self.enable_eStop_button.pressed.connect(self.enable_eStop)
+        self.enable_eStop_button.setEnabled(False)
+
+        self.disable_eStop_button = QPushButton('Reset Emergency Stop')
+        self.disable_eStop_button.pressed.connect(self.disable_eStop)
+        self.disable_eStop_button.setEnabled(False)
+
+        self.h_layout.addWidget(self.enable_eStop_button)
+        self.h_layout.addWidget(self.disable_eStop_button)
+
+        self.v_layout.addLayout(self.h_layout)
+
+        self.setLayout(self.v_layout)
+
+    def ping_srv(self, period):
+        if rospy.is_shutdown():
+            return False
+        try:
+            self.eStop_srv.wait_for_service(timeout=rospy.Duration(3.0))
+            if not self.eStop_srv_connected:
+                rospy.loginfo_throttle(1,'Connected to {}.'.format(self.eStop_srv_name))
+            self.eStop_srv_connected = True
+            return True
+        except:
+            rospy.loginfo_throttle(1,
+                                   "Lost Connection to {}".format(self.eStop_srv_name) if self.eStop_srv_connected else "Failed to connect to {}".format(self.eStop_srv_name))
+            self.eStop_srv_connected = False
+
+        
+        self.layout_estop_enable_btn(False)
+        self.layout_estop_disable_btn(False)
+
+    def enable_eStop(self):
```

Comment:
There is potentially a race condition here. What can happen here (although very unlikely(!)) is that the enable* and disable* are called one after another, you could have enable executing trigger service, while disable_eStop changes the self.eStop_req.data that it uses.

If you wanted to make sure this cannot happen, you could add a parameter to disable_eStop that will take this bool and then construct a SetBool request there instead of keeping it global.